### ISSUE TITLE
Change the Width of Discussion Board Text Box in SpeedGrading

### ISSUE THREAD

**Issue by martinmts** [jamesjonesmath/canvancement#40]
Hi,
First, thank you so much for developing and sharing these user scripts. These are extremely helpful!

I'd like to ask if you could help out with changing the width of the DB text book in the grading mode. As shown below, in the Elements tab of my Safari browser, I've IDed the code, as highlighted.

![image](https://user-images.githubusercontent.com/32575091/124946328-f15af000-dfdc-11eb-8d11-34038f1073ff.png)

This is a Zoom in look of the code sectoin:
![image](https://user-images.githubusercontent.com/32575091/124947255-ad1c1f80-dfdd-11eb-93a0-eb44f1f11e30.png)

My goal is to change the width, say, from '600px' to '800px'. This is my code:
>var div = document.querySelector("#content > div");
>div.style.width = "800px";

The problem I'm encountering: 
I can successfully make the change when I paste this code to the console of the browser. However, when I make it into a userscript, it won't work. And I found out that when I go to the "Sources" tab of Safari, the HTML code is different from the code in the "Elements" tab above. I've also located the parent `div` whose id = "`content`". However, what follows is different! It is no longer the `div` whose "style → width" = "600px".

See the screenshot below:

![image](https://user-images.githubusercontent.com/32575091/124946485-0f285500-dfdd-11eb-9c55-bcce6aa636c3.png)

I'm very puzzled and cannot understand why the code in the "Elements" tab, which accurately reflects the content of the webpage, is different from the "Sources" page. And the userscript seems to be working on the "Sources" page when I enter the speedgrader mode. (I believe this probably explains why the userscript does not work.)

I tried to search for '600px' in the code of the "Sources" tab, I couldn't find it.
![image](https://user-images.githubusercontent.com/32575091/124948675-e6a15a80-dfde-11eb-97bc-466ab202ffa7.png)

It would be extremely helpful if you could make a userscript or point me to a direction for a fix.

Thank you very much!

**Comment by jamesjonesmath** [comment]
Try using CSS rather than JavaScript and adding the !important attribute to it to override the inline style. JavaScript will only operate on the element if it is present, but CSS will persist.

Something like 
`#content > div { width: 100% !important; }`

The problem with this is is that you only want it to run in SpeedGrader

Adding a qualifier on the body element can help, but it probably won't work for you. We're now looking at 
`body.is-inside-submission-frame.no-headers   #content > div { width: 100% !important; }`

That may still be too generic as it will work on anything that is inside the SpeedGrader, not just discussions, but is is probably okay.

I took my [Canvas CSS Tweaks](https://github.com/jamesjonesmath/canvancement/blob/master/miscellaneous/canvas-css-tweaks.user.js) script, removed the code that checks to see if it's in an iframe, and then appended the lengthier rule and it worked.

If you want to write a script to do this, you can strip down that script and have it just run on
 `https://*.instructure.com/courses/*/assignments/*/submissions/*`

Realize that discussions are changing, so your CSS selector may change when they do.

**Comment by martinmts** [comment]
Hi James,
Thank you very much for your reply and offer sample codes.
I self-taught coding. So what I'm doing might be wrong. But here is part of the user script I have now:

>// @match		https://\*.instructure.com/courses/\*/gradebook/speed_grader\*
// ==/UserScript==
(function(){
	'use strict';
// common function in css to change css style
function addGlobalStyle(css) {
    var head, style;
    head = document.getElementsByTagName('head')[0];
    if (!head) { return; }
    style = document.createElement('style');
    style.type = 'text/css';
    style.innerHTML = css;
    head.appendChild(style);
}
addGlobalStyle('#content > div { width: 100% !important; }');

It's still not working.
I may have not successfully described the issue clearly. Let me try again:
 
The reason, I think, is not whether I'm using JavaScript or change CSS. When I copy and paste the code directly to the console of the browser, it works. It does not work when I use it as user script. I think the reason is that the code in the "Sources" tab is different from the code in the "Elements" tab, as I have shown in the OP.
The code in the "Elements" tab matches the content of the webpage, but the code in the "Sources" page does not. 

In the "Elements" tab, the line after the `id="content"` line is as follows:
`<div style="width: 600px; margin: 10px auto;">`
Screenshot:
![image](https://user-images.githubusercontent.com/32575091/125015134-e33cbc00-e03c-11eb-8aec-4c69504f513c.png)

I can then attempt to change the width of the `div` element.

However, in the "Sources" tab, the line after the `id="content"` line is as follows:
`<div id="speed_grader_timeout_alert"></div>`
Screenshot:
![image](https://user-images.githubusercontent.com/32575091/125015085-cdc79200-e03c-11eb-8177-73cc9d17cb56.png)

The user script sees to be applied to the code in the "Sources" tab, and if I copy and paste the JavaScript/CSS code to the Console of the browser, it is applied to the code in the "Element" tab. I think that explains why the direct paste to the console works and the user script does not work.

Thank you very much for helping out!

**Comment by jamesjonesmath** [comment]
I tested what I wrote before I replied.

Inject this line of CSS.

`body.is-inside-submission-frame.no-headers #content > div { width: 100% !important; }`

You cannot run it on the SpeedGrader URL because you're trying to modify the code within an iframe. The SpeedGrader URL is the surrounding material and controls. You need to match the src of the iframe.  That is, your match line needs to be 

`//@ match https://*.instructure.com/courses/*/assignments/*/submissions/*`

If your Canvas instance isn't at instructure.com, then you'll need to modify that portion of the match as well.

**Comment by martinmts** [comment]
Hi James,
Thank you so much for your patience. I'm starting to get what you meant by `iframe`. It works now. Learned something new today! 

May I ask what is `body.is-inside-submission-frame.no-headers` (I'm assuming this is the "qualifier" you said in the first reply)? I can see it is because of the iframe that is used. But I still have no clue where this whole string comes from. Googling it returns nothing.

Also, I have another quiz page that I want to change some style.
I can see that it is also inside a frame:
![image](https://user-images.githubusercontent.com/32575091/125149116-09c62a00-e105-11eb-9423-53e0d2975572.png)

I want to change, for instance, the style of the match questions, the answer submitted by students, and the answer key. 

Match question:
![image](https://user-images.githubusercontent.com/32575091/125149156-45f98a80-e105-11eb-9648-17b20858108e.png)

Answer:
![image](https://user-images.githubusercontent.com/32575091/125149164-5f023b80-e105-11eb-8c21-becfcb5e06cc.png)

Alternative answers:
![image](https://user-images.githubusercontent.com/32575091/125149175-6d505780-e105-11eb-944f-8ef24d028383.png)

I used the following code, but it does not work.
```
// @match		https://*.instructure.com/courses/*/assignments/*/submissions/*
// ==/UserScript==

(function(){
	'use strict';

	function addGlobalStyle(css) {
		var head, style;
		head = document.getElementsByTagName('head')[0];
		if (!head) { return; }
		style = document.createElement('style');
		style.innerHTML = css;
		head.appendChild(style);
	}

	addGlobalStyle('body.is-inside-submission-frame.no-headers .answer_match_left { font-family: "Times New Roman" !important; font-size: 30px !important; }');
	addGlobalStyle('body.is-inside-submission-frame.no-headers .answer { font-family: "Times New Roman" !important; font-size: 30px !important; }');
	addGlobalStyle('body.is-inside-submission-frame.no-headers .answer_text { font-family: "Times New Roman" !important; font-size: 30px !important; }');	
})();
```

Can you take a look at it when you have time?
Thank you very much!

**Comment by jamesjonesmath** [comment]
The `body.is-inside-submission-frame.no-headers` is a CSS selector that says you must match the body element that has classes of _is-inside-submission-frame_ and _no-headers_. That's an attempt to narrow down the pages that it runs on. If you're using it within a UserScript where you have a match statement, it may not be important. I was originally trying to write something that could also be put into the global custom JavaScript, where it might be applicable.

Your code above creates 3 stylesheets. That's not necessarily terrible, but it's not efficient either. I would recommend that you take a look at the Canvas CSS Tweaks I recommended in the original reply. It creates one stylesheet that holds all of the rules. There is a more efficient way (fewer lines) of adding stylesheets to a page.

Understanding the purpose of the specific selector goes a long way towards explaining why your new code doesn't work. When you look at the body element for a quiz, it doesn't have _is-inside-submission-frame_, it has _quizzes-speedgrader_ as a class on the body element. I'm not even going to check the specific CSS you're injecting because it's not getting ran because of your selectors.

You probably shouldn't be using !important for everything. I included it because I was overriding inline styles and !important is the only way to do that with a stylesheet. The [MDN page on Specificity](https://developer.mozilla.org/en-US/docs/Web/CSS/Specificity) says using it is bad practice and you should always look for a way to use specificity before considering !important. There are no inline styles for what you have, that means the you should really be more specific rather than using !important. Adding an element before that (like I did with the body) increases the specificity, but it could also be a `#content` or any other parent element. The more classes or IDs or elements you throw in, the more specific it is and there's no need to override the code with !important. Note that adding !important won't fix things if the selector doesn't match, which is what is happening in your case.

**Comment by martinmts** [comment]
Hi James,
Thanks again for your detailed explanation. I think I can understand what you are saying now.
I did look at your Canvas CSS Tweaks, but at the first glance, I was overwhelmed by the other options that were offered there, and I did not know that to edit. With more understanding and after a closer reading, it appears (if I'm understanding correctly), I only need to change three places:
1. the `@match` line, which I changed to `// @match       https://*.instructure.com/courses/*/assignments/*/submissions/*`
2. comment out the `iframe` check.
```
/*
  if (window !== parent.window) {
    return;
  }
*/
```
3. add rules.
I added only one rule for test:
```
  const rules = ['div.answer_match_left { font-family: "Times New Roman"; font-size: 30px; }'];
```
I assume multiple rules will be as such:?
```
  const rules = ['.answer_match_left { font-family: "Times New Roman"; font-size: 30px; }',
                 '.answer { font-family: "Times New Roman"; font-size: 30px; }',
                 '.answer_text { font-family: "Times New Roman"; font-size: 30px; }'
               ];
```
I'm can understand what you are saying about specific selector now. I actually also tested the code with `body.is-inside-submission-frame.no-headers` removed. So, the selector was `.answer_match_left` only. However, it did not work either.

 Now, this is my code but it is not working:
```
// ==UserScript==
// @name        Canvas CSS Tweaks
// @namespace   https://github.com/jamesjonesmath/canvancement
// @description Various tweaks to Canvas CSS
// @match       https://*.instructure.com/courses/*/assignments/*/submissions/*
// @version     5
// ==/UserScript==
(function () {
  'use strict';

  // Only run in parent window, do not run in iframes
/*
  if (window !== parent.window) {
    return;
  }
*/

  // Select the features to enable. Comment out or delete any you do not want
  const features = [
    'fullscreen', // use full width of page
    'sg_rubrics', // don't make rubrics so big in speedgrader
    'sg_avatar_zoom', // zoom avatar in SpeedGrader
    // 'sg_long_descriptions', // hide long descriptions in SpeedGrader rubrics
    // 'sg_points_possible', // hide criterion points possible in SpeedGrader rubrics
    // 'no_test' , // hide test instance notification in test and beta
  ];

  const rules = ['div.answer_match_left { font-family: "Times New Roman"; font-size: 30px; }'];

// some lines here have been omitted to keep the code here brief.

  if (rules.length) {
    const style = document.createElement('style');
    document.head.appendChild(style);
    const sheet = style.sheet;
    for (let i = 0; i < rules.length; i++) {
      sheet.insertRule(rules[i],i);
    }
  }

}) ();

```

I'm very thankful for your prompt and detailed explanation. It is my first time to modify CSS via JavaScript. I appreciate your patience, detailed explanation, and even going all the way to repeat what you already said. I got the code I was using from other places. Now I'm trying to use your Canvas CSS Tweaks. If you could take a look at the above code I modified based on your template, and let me know what I'm doing wrong, it'll be great!

The relevant code sections:
iframe:
![image](https://user-images.githubusercontent.com/32575091/125169770-b04c1280-e179-11eb-97a8-7aaf28d79307.png)

Match questions:
![image](https://user-images.githubusercontent.com/32575091/125169786-c2c64c00-e179-11eb-803f-787946ea5c04.png)

I guess I could also add the `body` and `div` to the selector like this:
```
const rules = ['body.quizzes-speedgrader.no-header div.answer_match_left { font-family: "Times New Roman"; font-size: 30px; }'];
```
![image](https://user-images.githubusercontent.com/32575091/125169900-57c94500-e17a-11eb-98c8-1128ce86bc9c.png)

However, since the more generic `.answer_match_left` did not work, I can't test a more specific selector.

Also, thank you for your explanation about `!important`. I will try to avoid it as possible.

**Comment by jamesjonesmath** [comment]
It turns out that the quizzes have a different source than the discussions, so the `/courses/*/assignments/*/submissions/*` won't work there. That submission gets redirected to a different page when you view a quiz, so it wasn't matching. A path that will match is `/courses/*/quizzes/*/history?*`. You can check whether the script is running at all by looking at the Tampermonkey icon on the toolbar. It has a red badge with the number of scripts that are running or click on the Tampermonkey script to get a list of scripts that are running on the page.

If you want one script that will work for both discussions and quizzes, you need to have multiple `// @match` lines in the metadata.

You can also specify multiple CSS selectors on a single line, separated by a comma. You can specify them separately like you are doing as well -- it's more readable and if you're just starting out, that may be more important than combining the rules. If you want to style them differently, you would need separate rules.

Taking what you've done and stripping out the extra stuff from my CSS tweaks script, I come up with something like this that works for both discussions and quizzes. I removed the link to my site so that it doesn't try to update based on my script.

```
// ==UserScript==
// @name        SpeedGrader Tweaks
// @description Various tweaks to SpeedGrader CSS
// @match       https://*.instructure.com/courses/*/assignments/*/submissions/*
// @match       https://*.instructure.com/courses/*/quizzes/*/history?*
// @version     1
// ==/UserScript==
(function() {
  'use strict';

  const rules = ['body.is-inside-submission-frame #content > div { width: 100% !important; }',
    'body.quizzes-speedgrader .answer_match_left, body.quizzes-speedgrader .answer, body.quizzes-speedgrader .answer_text { font-family: "Times New Roman"; font-size: 30px; }',
  ];

  if (rules.length) {
    const style = document.createElement('style');
    document.head.appendChild(style);
    const sheet = style.sheet;
    for (let i = 0; i < rules.length; i++) {
      sheet.insertRule(rules[i], i);
    }
  }

})();
```

The rule with the body.quizzes-speedgrader repeated makes for a long line. The .answer classes are unlikely to appear on the discussions page, so you could probably leave them out. I have not tested that, though, nor looked through the source code. I like to future-proof things, even though Quizzes is being deprecated, so I included them here. 

Just using `'.answer_match_left, .answer, .answer_text { font-family: "Times New Roman"; font-size: 30px; }'` seems to work, but it has the same specificity as the CSS that Canvas uses, so it may not guaranteed to always work. You could preface each of those with a `div` to increase the specificity so it looks like `'div.answer_match_left, div.answer, div.answer_text { font-family: "Times New Roman"; font-size: 30px; }'`.

I did remove the _.no-headers_ from rule for the discussion. I only had that on there to help make sure it would work if included in the global custom CSS for all users. This really looks like a user customization rather than something you're trying to do for all users, though, so it doesn't have to be as specific.

**Comment by martinmts** [comment]
Hi James,
It's working now!
I can finally have a BIG smile on my face! Thank you so much!!

Just one small question about what you said:
>It turns out that the quizzes have a different source than the discussions, so the /courses/*/assignments/*/submissions/* won't work there. That submission gets redirected to a different page when you view a quiz, so it wasn't matching. A path that will match is `/courses/*/quizzes/*/history?*`.

I searched for `<iframe` and found only three instances as follows:
![image](https://user-images.githubusercontent.com/32575091/125173725-a680da00-e18e-11eb-8296-bed4ed1ccac0.png)

Only the middle one has `/courses/*/assignments/*/submissions*`
The other two are `about:blanks`.

How did you find out `/courses/*/quizzes/*/history*` is what we are looking for?

I did find one instance of the link pattern, but it is inside a `<script>` label and I would never know this is what I need if you did not tell me.
![image](https://user-images.githubusercontent.com/32575091/125173839-59e9ce80-e18f-11eb-8467-ef01648f1478.png)


>You can check whether the script is running at all by looking at the Tampermonkey icon on the toolbar. It has a red badge with the number of scripts that are running or click on the Tampermonkey script to get a list of scripts that are running on the page.
Thank you very much for the tip. I use the Safari browser for grading work, but I did not purchase the Tampermonkey plugin. I have Adguard, which can manage user scripts at one place and work with multiple browsers. The downside is that it does not tell whether the user script is running or not. But this is a very useful tip. Next time, I will test it with another browser that has Tampermonkey installed.

**Comment by jamesjonesmath** [comment]
To find the source of the content, I looked at the network tab of Chrome's developer tools, cleared out all the logs so I could tell what was new, then moved from one student to another to see what was being loaded in the iframe. The first one is often the one you're after as it is the main page. Make sure you are showing _All_ and not just _XHR_ requests. Alternatively, you can specify _Doc_ as the type and it just lists documents.

Sometimes you need to preserve (Chrome) or persist (Firefox) the logs, especially with a POST. In this case, it's just a GET. From there, you can get the URL from the Headers tab.

Another benefit is that I had done a lot of work with SpeedGrader before when I wrote QuizWiz and so it started coming back to me, but other people don't have that experience to fall back on.

Here's the first request when I moved to a new student. Notice the first one has a 302 status (redirect) while the second one has a 200 status (OK).

![image](https://user-images.githubusercontent.com/14840711/125174607-34a79100-e18c-11eb-9110-0c9fa87a2014.png)

You cannot get the full URL from the summary screen, you need to look at the Headers tab. Note that the first one matches `/courses/*/assignments/*/submissions/*` pattern but had a 302 status.

![image](https://user-images.githubusercontent.com/14840711/125174537-adf2b400-e18b-11eb-9b54-23c7ce594e31.png)

The second one matches the `/courses/*/quizzes/*/history` path with a 200 status.

![image](https://user-images.githubusercontent.com/14840711/125174516-9a474d80-e18b-11eb-812c-00fc5d78469c.png)

Tampermonkey is matching against the entire URL, not just the path, so I added `?*` but I could have also added `?headless=1&user_id=*&version=*` instead. There's no guarantee that the order of the query parameters will be consistent, though, so I stopped after the `?`. I could have also used `/history*` but that would have matched `/historyMartin`. Although Canvas doesn't use `historyMartin`, I don't know that it doesn't use some other word starting with history.

**Comment by martinmts** [comment]
Wow~I knew nothing about what the "Network" tab was for.
Thank you so much for showing all of this. Safari has slightly different arrangement, but I can see both the "assignments" link and the "quizzes" link in the Summary URLs with the Status 200.
![image](https://user-images.githubusercontent.com/32575091/125175401-d1bcf680-e199-11eb-85aa-bea83cb70531.png)

Your explanations have been enormously helpful! You not only solved my problem but also taught me how to solve it.
Thank you so much and have a great the rest of the day!
